### ISSUE TITLE
Array as inputs and outputs using Input method

### ISSUE THREAD

**Issue by simonMoisselin** [replicate/cog#527]
Hello,

Is it possible to add the type `array` as an input, in addition to images or files?

**Comment by zeke** [comment]
Hi @simonMoisselin 👋🏼 

Arrays/lists as inputs and outputs are not currently supported in Cog, but this seems to me like a reasonable thing to support. Can you share a little about your intended use case?

For a list of types that are currently supported, see the docs on [Input and output types](https://github.com/replicate/cog/blob/684c735335d8a51676ab6b611d8535572ae8dd93/docs/python.md#input-and-output-types).

See also this open issue about supporting multiple file outputs: https://github.com/replicate/cog/issues/411

**Comment by dashstander** [comment]
I think this is a great idea. I don't know necessarily if this would work the best (in particular it might be somewhat extraneous to using models from replicate.com), but would be awesome to support [Arrow](https://arrow.apache.org/docs/index.html) for serializing and de-serializing array data.

**Comment by anotherjesse** [comment]
This is still the case - we do not yet support lists/arrays of input:

```
TypeError: Unsupported input type typing.List[float] for parameter `nums`. Supported input types are: str, int, float, bool, cog.File, cog.Path.
ⅹ Failed to get container status: exit status 1
```

**Comment by jayaffine** [comment]
This would be a great idea to support different formats.
I wanted to host some derivatives of Segment Anything Model  where the results  are in `dict` format.
So the results will have numpy arrays, boolean arrays, etc.
It would be great if I can pass these dict files as results.

> Hi @simonMoisselin 👋🏼
> 
> Arrays/lists as inputs and outputs are not currently supported in Cog, but this seems to me like a reasonable thing to support. Can you share a little about your intended use case?
> 
> For a list of types that are currently supported, see the docs on [Input and output types](https://github.com/replicate/cog/blob/684c735335d8a51676ab6b611d8535572ae8dd93/docs/python.md#input-and-output-types).
> 
> See also this open issue about supporting multiple file outputs: #411

**Comment by bfirsh** [comment]
Related / dupe of https://github.com/replicate/cog/issues/608

**Comment by alexblattner** [comment]
hi, I'd like to be able to upload multiple pictures for a single argument. This would be very useful for that

**Comment by camenduru** [comment]
Hi @zeke 👋 Dust3r ( https://replicate.com/camenduru/dust3r ) needs multiple images as input.

**Comment by yosun** [comment]
@zeke this is a bug not an enhancement! i am afraid to upload pictures of corgi.cam but, if you want to know, currently, with only 2 photos, this issue is killing pugs: https://replicate.com/p/llpqfprbx6oo3hmnggl4nbvqku

**Comment by yosun** [comment]
[![Image from Gyazo](https://i.gyazo.com/e89f1c260b4d452fcba03f95987a9f14.gif)](https://gyazo.com/e89f1c260b4d452fcba03f95987a9f14)

**Comment by zeke** [comment]
Raised it with the team. We're working on it!

**Comment by dkhokhlov** [comment]
the issue with List[str] type is confirmed. temp workaround - use simple 'str' type and split value later.
looking into it.

**Comment by zeke** [comment]
>  use simple 'str' type and split value later.

Can you provide an example? How would that work for files?

**Comment by yosun** [comment]
> the issue with List[str] type is confirmed. temp workaround - use simple 'str' type and split value later. looking into it.
> 
> $ cog --version cog version 0.9.4 (built 2024-01-24T22:16:49Z)

okay so basically we can just concatenate as many input URLs as we want and then parse split it?

this does not seem secure. kind of like injection attack looming to happen.

**Comment by dkhokhlov** [comment]
right, looking into it.

**Comment by dkhokhlov** [comment]
@yosun 

cog on "support_for_list_in_input" PR branch supports list[File] input:

```
 $ cat predict.py
from cog import BasePredictor, File

class Predictor(BasePredictor):
   def predict(self, files: list[File]) -> str:
       output_parts = []  # Use a list to collect file contents
       for f in files:
           # Assuming file content is in bytes, decode to str before appending
           content = f.read()
           output_parts.append(f"{f} size={len(content)} bytes")
       return "\n".join(output_parts)

$ cog predict -i 'files=["https://replicate.delivery/pbxt/KVsSoflaESRsavl3dTJ3cAUggGPnMlXZnEEi70Sll8tvfByH/Pug%20Dog.H14.2k.png", "https://replicate.delivery/pbxt/KVsSoQePGoqo1fUwrXBnEbxsDuKkDAMKlk6mFZnO5McoRv8F/Pug%20Dog.H03.2k.png"]'

Running prediction...
<URLFile at 0x7fee0efb5300 wrapping <urllib3.response.HTTPResponse object at 0x7fee0efa9b70>> size=5123929 bytes
<URLFile at 0x7f9173aa0ac0 wrapping <urllib3.response.HTTPResponse object at 0x7f9173a94e80>> size=5122833 bytes
```
note - list[File] and how it handled. is it ok? 
or you prefer Path instead of File?

**Comment by dkhokhlov** [comment]
Ref: 
https://linear.app/replicate/issue/PRO-1153/add-support-for-list[file]-input-to-web-ui

**Comment by simonMoisselin** [comment]
> @yosun
> 
> cog on "[support_for_list_in_input" PR branch](https://github.com/replicate/cog/pull/1561) supports list[Path] input. For local runs only, not supported in Web UI yet, in progress. Example:
> 
> ```
> $ echo test1 > 1.txt
> $ echo test2 > 2.txt
> $ cat predict.py
> from cog import BasePredictor, Path
> 
> class Predictor(BasePredictor):
>    def predict(self, paths: list[Path]) -> str:
>        output_parts = []  # Use a list to collect file contents
>        for path in paths:
>            with open(path) as f:
>              output_parts.append(f.read())
>        return "".join(output_parts)
> 
> $ cog predict -i 'paths=["@1.txt", "@1.txt"]'
> 
> Running prediction...
> test1
> 
> test2
> ```

Thanks a lot @dkhokhlov!

**Comment by yosun** [comment]
is this also supported via API, if so what's the brief format?

**Comment by yosun** [comment]
> @yosun
> 
> cog on "[support_for_list_in_input" PR branch](https://github.com/replicate/cog/pull/1561) supports list[Path] input. For local runs only, not supported in Web UI yet, in progress. Example:
> 
> ```
> $ echo test1 > 1.txt
> $ echo test2 > 2.txt
> $ cat predict.py
> from cog import BasePredictor, Path
> 
> class Predictor(BasePredictor):
>    def predict(self, paths: list[Path]) -> str:
>        output_parts = []  # Use a list to collect file contents
>        for path in paths:
>            with open(path) as f:
>              output_parts.append(f.read())
>        return "".join(output_parts)
> 
> $ cog predict -i 'paths=["@1.txt", "@1.txt"]'
> 
> Running prediction...
> test1
> 
> test2
> ```

is this also supported via API, if so what's the brief format?

**Comment by dkhokhlov** [comment]
yes, it will be supported in api, after new cog is released. it is transparent for api.

**Comment by dkhokhlov** [comment]
it will be supported by api. format should  be the same as with list[str] case, except str is uri str now..

**Comment by markphelps** [comment]
This is a duplicate of #608, which tracks support for lists/arrays as inputs and outputs. I'm closing this issue to consolidate discussion there—please follow #608 for updates.
